### DISCUSSION TITLE
I am trying to deserialize canvas with custom class in angular

### DISCUSSION THREAD

**Discussion by zmiju** [fabricjs/fabric.js#8631] [Q&A] [answered]
I have found examples saying to write short function:
```
fabric.labeledRect.fromObject = function(object, callback) {
  return fabric.Object._fromObject('LabeledRect', object, callback);
}
```

But I am using angular with typescript and after spending some hours I have no idea how to implement it in angular/typescript. Please help.

I tried something like this, but it didn't work:
```
  TextboxWithPadding = fabric.util.createClass(fabric.Textbox, {
    type: 'TextboxWithPadding',
    toObject: function() {
      return fabric.util.object.extend(this.callSuper('toObject'), {
        backgroundColor: this.get('backgroundColor'),
        padding: this.get('padding'),
        cornerRadius: this.get('cornerRadius')
      });
    },
     fromObject: function(object: any, callback: any) {
            return fabric.Object._fromObject('TextboxWithPadding', object, callback);
        },
  
    _renderBackground: function(ctx: any) {
      if (!this.backgroundColor) {
        return;
      }
      var dim = this._getNonTransformedDimensions();
      ctx.fillStyle = this.backgroundColor;
      ctx.beginPath();
      ctx.roundRect(-dim.x / 2 - this.padding, -dim.y / 2 - this.padding, dim.x + this.padding * 2, dim.y + this.padding * 2, [this.cornerRadius]); 
      ctx.fill()
      ctx.stroke()
      this._removeShadow(ctx); 
    }
  });
```
I also tried to put fromObject like this, but it also didn't work:
```
       this.TextboxWithPadding.fromObject = function(object: any, callback: any) {
          return fabric.Object._fromObject('TextboxWithPadding', object, callback);
        }
```

**Comment by ShaMan123** [comment, marked as answer]
All in all it seems you are on the right track.
Custom objects are not documented and awkward, hopefully moving to classes will make it easier for us all.
My advice is for v5 and under since v6 it different thankfully.
Did you assign the class to fabric global object?
```
fabric.TextboxWithPadding = TextboxWithPadding
```
Without that calling fromObject won't work becuase it looks in the fabric global object for the class and tries to resolve it.

Also might be that type is wrong
```diff
type: 'textboxWithPadding'
```

**Reply by ShaMan123** [reply]
future proofing your code for v6 I suggest you avoid using `fabric.util.object.XXX` as well as `this.get`

**Reply by zmiju** [reply]
Aaaaa, now I understand the logic :)
I didn't do it, but the reason is that it yields an errror message: [{
`...Property 'TextboxWithPadding' does not exist on type 'typeof import(\"/Users/marcinzmigrodzki/Documents/angularapps/ll/node_modules/@types/fabric/fabric-impl\")'."...`

**Reply by ShaMan123** [reply]
yes, that types package will die as well in v6, it is inaccurate for any version
you can declare the class

**Reply by zmiju** [reply]
Thanks for comment. However my skills are to low to derive proper code from your comment. I would be more than happy if anyone provided me an example code.

**Reply by ShaMan123** [reply]
It works the other way around.
You provide a repro so someone can suggest a solution in a few minutes.

**Reply by zmiju** [reply]
Alright, it is the first time for me to discuss issues on Github. Thanks for remark.
Could you tell me how to provide repro?

**Reply by ShaMan123** [reply]
When submitting a [new issue](https://github.com/fabricjs/fabric.js/issues/new?assignees=&labels=&template=bug_report.yml&title=%5BBug%5D%3A+) there are links under the `Link To Reproduction` section

**Reply by zmiju** [reply]
I created a stackblitz example: https://angular-mzoqtk.stackblitz.io

**Reply by ShaMan123** [reply]
seems to work

**Reply by zmiju** [reply]
If "seems to work" you mean that my example shows the bug, you are right.
But if "seems to work" means that there is no problem in the code, you are wrong. I cannot load json and when I try to add this:
`fabric.TextboxWithPadding = TextboxWithPadding`
I am getting error

**Reply by zmiju** [reply]
look at errror in developer's console after clicking button "LOAD"

**Reply by ShaMan123** [reply]
I will instruct you a bit more...
It is useless sharing a repro without having decent access to the code

**Reply by zmiju** [reply]
You mean this : https://stackblitz.com/edit/angular-mzoqtk?file=src%2Fmain.ts

**Reply by ShaMan123** [reply]
yes
don't let people helping you guess how to help you
That is a bad idea

**Reply by ShaMan123** [reply]
Just pointing out that once you put the proper link to the live code it took me less than a minute to help you...
Help others help you!
https://stackblitz.com/edit/angular-mvcveu?file=src%2Fmain.ts

**Reply by zmiju** [reply]
Sorry, it is the first time I submit anything to git and the first time I created stackblitz. But I am learning ;)

**Reply by ShaMan123** [reply]
Great
Welcome to fabric!
The problem was:
- typescript strict config in your app (it is supposed to help you and instead it bugs you, a known thing)
- Text classes are initialized with a text value as first arg so `_fromObject` needs to take that into consideration with `extraParam` config

**Reply by zmiju** [reply]
Wow!!!!
Whole day lost, and you solved it in 1 minute.
Thanks a lot!

**Reply by ShaMan123** [reply]
Sorry for that.
I know how frustrating fabric can be for newbies. It is a steep learning curve.
Docs are horrible.

**Reply by ShaMan123** [reply]
Don't be afraid looking into the source code. Now that it is written in TS you can actually navigate through it and get a feeling of what you should tryout

**Reply by zmiju** [reply]
Once again, many thanks for help :)

**Comment by ShaMan123** [comment]
BTW since you are starting I would flag the fact that we are on the verge of publishing v6 which is far better in structure and in features than v5.
Consider being an early adopter but that comes with cons as well as pros.
#8299

**Reply by zmiju** [reply]
I will check this new version out since I am starting to develop new solution based on fabric.

**Reply by ShaMan123** [reply]
We have a few PRs that must merge to fix some stuff and then I guess we will announce beta
